scipy/stats/mstats_extras.py

```diff
@@ -161,13 +161,12 @@ def _hdsd_1D(data, prob):
 
         vv = np.arange(n) / float(n-1)
         betacdf = beta.cdf
-
+        
         for (i,p) in enumerate(prob):
             _w = betacdf(vv, (n+1)*p, (n+1)*(1-p))
             w = _w[1:] - _w[:-1]
-            mx_ = np.fromiter([np.dot(w,xsorted[np.r_[list(range(0,k)),
-                                                      list(range(k+1,n))].astype(int_)])
-                                  for k in range(n)], dtype=float_)
+            mx_ = np.fromiter([np.dot(w, np.delete(xsorted, k))
```

Comment:
```suggestion
            mx_ = np.fromiter((np.dot(w, np.delete(xsorted, k))
```
consider turning the list comprehension into a generator expression (i.e. `[` -> `(` and `]` -> `)`) to avoid forming the list explicitly then copying it into an array.

Reply:
I like this idea, `w[:k] @ xsorted[:k] + w[k:] @ xsorted[k+1:]`, very elegant. It's faster too (see below).

I had already tried to use a generator expression, but creating a local function for the indexing, and it was a lot slower and had problems with "Can't pickle this expression" errors. With your version, it's easy to go from list comprehension to generator. I've created new speed tests, and I'll put them below.

Reply:
From the tests, the list comprehension is slightly faster than the generator expression, so I've commited that one